README.md
#
# Cookbook Name:: harden_macos
# Author:: Meg Cassidy (<meg@nuna.com>)
#
# Copyright:: 2016-2017, Nuna, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# harden_macos Cookbook
This cookbook covers several areas of hardening for OS X and has generally 
only been tested for OSX version 10.11.x and 10.12.x.

## Requirements
### platforms
- `mac_os_x`

### platform_version
- `10.11.x`
- `10.12.x`

### cookbooks
- `mac_os_x` - For userdefaults custom resource to do Library defaults write commands

## Attributes
TODO: fix this
This cookbook does not currently utilize attributes.

## References
### GOV.UK
- https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/470580/osx-provisioning-script.sh.txt - Specific 10.11 commands
- https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/471620/End_User_Devices_Security_Guidance_-_Apple_OS_X_10_11.pdf - Security explanations

###  Apple
- https://support.apple.com/en-us/HT201159 (links to gov.uk provisioning script)

### Other
- https://github.com/cagerton/dots/blob/master/macos.sh - Specific MacOS commands

## Usage
The default recipe is meant to call the other parts of the recipes. Each of
the recipes consist of different functional groups of checks and changes.
These groups are explained below:

### harden_macos::default
TODO: fill in

Just include `harden_macos::default` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[osquery]"
  ]
}
```

## License and Authors
Authors: Meg Cassidy (meg@nuna.com)


CHANGELOG.md
# harden_macos cookbook CHANGELOG

This file is used to list changes made in each version of the harden_macos cookbook.

1.0.0
-----
- Meg - Initial release of harden_macos cookbook to open source.


- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: harden_macos
# Author:: Meg Cassidy (<meg@nuna.com>)
# Spec:: default
#
# Copyright:: 2016-2017, Nuna, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'spec_helper'

describe_recipe 'harden_macos::default' do
  context 'When running on mac_os_x 10.11' do
    let(:node_attributes) do
      { platform: 'mac_os_x', version: '10.11.1' }
    end

    # let(:context_node_attributes) do
    #   os_version == 11
    # end

    it 'converges successfully' do
       expect { chef_run }.to_not raise_error
    end

    it 'runs harden task with the default action' do
      expect(chef_run).to secure_harden_macos('firewall')
    end
    # it 'runs the harden_os updates topic' do
    #   expect(chef_run).to run_harden_macos('updates')
    # end

    # https://github.com/sethvargo/chefspec#testing-lwrps
  end
end


spec/unit/recipes/updates_spec.rb
#
# Cookbook Name:: harden_macos
# Author:: Meg Cassidy (<meg@nuna.com>)
# Spec:: updates
#
# Copyright:: 2016-2017, Nuna, Inc.
#
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# require 'spec_helper'

# describe 'harden_macos::updates' do
#   context 'When running on mac_os_x' do
#     let(:chef_run) do
#       s = ChefSpec::SoloRunner.new(platform: 'mac_os_x', version: '10.11.1')
#       s.converge(described_recipe)
#     end

#     it 'converges successfully' do
#       expect { chef_run }.to_not raise_error
#     end

#     before do
#       stub_command('softwareupdate --schedule | grep "Automatic check is off"').and_return(true)
#     end

#     it 'executes softwareupdate schedule on command' do
#       expect(chef_run).to run_execute('softwareupdate on')
#     end

#   end
# end



test/integration/default/serverspec/hardening_spec.rb
#
# Cookbook Name:: harden_macos
# Author:: Meg Cassidy (<meg@nuna.com>)
#
# Copyright:: 2016-2017, Nuna, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'spec_helper'

describe 'harden_macos::default' do
  # Serverspec examples can be found at
  # http://serverspec.org/resource_types.html
  describe 'enable firewall' do
    describe command('defaults read /Library/Preferences/com.apple.alf globalstate') do
      its(:stdout) { should match /^1$/ }
    end

    describe 'enable firewall stealthmode' do
      describe command('/usr/libexec/ApplicationFirewall/socketfilterfw --getstealthmode') do
        its(:stdout) { should match /Stealth mode enabled/ }
      end
    end
  end

  describe 'Safari disable opening files after downloading' do
    describe command('defaults read /Users/vagrant/Library/Preferences/com.apple.Safari AutoOpenSafeDownloads') do
      its(:stdout) { should match /^0$/ }
    end
  end

  describe 'Disable Safari Spotlight Suggestions' do
    describe command('defaults read /Users/vagrant/Library/Preferences/com.apple.Safari UniversalSearchEnabled') do
      its(:stdout) { should match /^0$/ }
    end
  end

end


resources/default.rb
#
# Cookbook Name:: harden_macos
# Author:: Meg Cassidy (<meg@nuna.com>)
# Resources:: default
#
# Copyright:: 2016-2017, Nuna, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

resource_name :harden_macos

actions :secure
default_action :secure

# A topic is a named group of hardening tasks
# TODO(meg): Include list of topics and their tasks in README
attribute :topic,
          kind_of: String,
          name_attribute: true,
          required: true

# Task specific options, such as setting the screensaver to 5 minutes, etc.
attribute :options,
          kind_of: Hash,
          required: false

# Run as a specific user, otherwise run as node's current user
attribute :user,
          kind_of: String,
          default: node['harden_os']['user'],
          required: true

action :secure do
  unless node['harden_os']['user'].nil?
    case node['platform']
    when 'mac_os_x'
      case new_resource.topic.downcase
      # Call specific methods for each topic
      when 'firewall' then harden_macos_firewall
      when 'safari' then harden_macos_app_safari
      when 'macos mail' then harden_macos_app_mail
      when 'devices' then harden_macos_devices
      when 'finder' then harden_macos_finder
      when 'icloud' then harden_macos_icloud
      when 'network' then harden_macos_network
      when 'remote services' then harden_macos_remote_services
      when 'sharing' then harden_macos_sysprefs_sharing
      when 'utilities' then harden_macos_utilities
      when 'ssh configs' then harden_macos_ssh_configs
      when 'updates' then configure_macos_updates
      when 'user preferences' then configure_macos_user_prefs
      when 'user permissions' then configure_macos_user_permissions
      when 'login window' then configure_macos_login_window_prefs
      when 'sleep' then configure_macos_sleep_prefs
      when 'hotcorners' then configure_macos_hotcorners
      when 'privacy' then configure_privacy_prefs
      end
    else
      # Do not extend this functionality to support other operating systems
      raise "Node platform (#{node['platform']}) is not supported at this time"
    end
  end
end


libraries/os_helpers.rb
#
# Cookbook Name:: harden_macos
# Library:: os_helpers
# Author:: Meg Cassidy (<meg@nuna.com>)
#
# Copyright:: 2016-2017, Nuna, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Helper methods for general OS related case statements
module OperatingSystemHelpers

  def os_version
    # Takes the second integer from the node['platform_version'] value to run 
    # checks against, since the cookbook only deals with macOS and that is used
    # as the major version.
    node['platform_version'].split('.')[1].to_i
  end

  def supported
    # Key/value array used in 'supported_platform_version' and 
    # 'is_supported_platform?' methods against the node['platform'] value.
    # TODO: make work with 10.12. Might need update to logic in supported_platform_version if it is looking at the whole version number? 
    {
      mac_os_x: %w(10.11)
    }
  end

  def supported_platform_version
    # Search the 'supported' method array contents for node['platform_version']
    # values and returns true or false
    current_version = Chef::Version.new(node['platform_version'])
    sp = false
    supported[node['platform'].to_sym].each do |version|
      required_version = Chef::Version.new(version)
      next unless required_version.major == current_version.major
      sp = true if required_version <= current_version
    end
    sp
  end

  def is_supported_platform?
    # Returns true or false if node['platform'] is contained in the 'supported'
    # method array AND the platform version is a value for that key
    supported.keys.include?(node['platform'].to_sym) &&
      supported_platform_version
  end
end

Chef::Recipe.send(:include, OperatingSystemHelpers)
Chef::Resource.send(:include, OperatingSystemHelpers)
Chef::Provider.send(:include, OperatingSystemHelpers)


resources/atypical_plist.rb
#
# Cookbook:: harden_macos
# Author:: Craig Anderson (<craig@nuna.com>)
# Provider:: atypical_plist
#
# Copyright:: 2016-2017, Nuna, Inc.
#
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Include the plist library file
include ReadWritePlist

actions :update
default_action :update

#TODO: Finish fixing this based on compatresource cookbook guidelines
property :identifier, :kind_of => String, :name_attribute => true, :required => true
property :values, :kind_of => [Hash], :default => nil, :required => true
property :changelist, :kind_of => [Hash], :default => nil, :required => false


libraries/harden_macos.rb
#
# Cookbook Name:: harden_macos
# Library:: harden_macos
# Author:: Meg Cassidy (<meg@nuna.com>)
#
# Copyright:: 2016-2017, Nuna, Inc.
#
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

module OSHardeningCookbookMacOS
  module HardeningHelpers

    def harden_macos_firewall
      puts("\nHardening Library: " + __method__.to_s)
      # Enable the OS X System Firewall
      #   1 = enable firewall and disables "Block All Incoming Connections"
      #   2 = enable firewall and enable the setting "Block All Incoming
      #   Connections."
      mac_os_x_userdefaults 'Enable OS X system firewall' do
        domain '/Library/Preferences/com.apple.alf'
        key 'globalstate'
        value 1
        only_if { File.exist? '/Library/Preferences/com.apple.alf.plist' }
      end

      execute 'Enable firewall stealthmode' do
        command '/usr/libexec/ApplicationFirewall/socketfilterfw --setstealthmode on'
        not_if '/usr/libexec/ApplicationFirewall/socketfilterfw '\
               '--getstealthmode | grep "Stealth mode enabled"'
      end
    end

    def harden_macos_app_safari
      puts("\nHardening Library: " + __method__.to_s)
      mac_os_x_userdefaults 'Safari disable opening files after downloading' do
        domain 'com.apple.Safari'
        key 'AutoOpenSafeDownloads'
        value 0
        user new_resource.user
      end

      # Safari's separate setting (separate from Spotlight) for search metrics
      mac_os_x_userdefaults 'Disable Safari Spotlight Suggestions' do
        domain 'com.apple.Safari'
        key 'UniversalSearchEnabled'
        value 0
        user new_resource.user
      end
    end

    def harden_macos_app_mail
      puts("\nHardening Library: " + __method__.to_s)
      # com.apple.mail-shared only exists setup of Mail.app was initiated
      mac_os_x_userdefaults 'Disable autoload remote content in Mail.app' do
        domain 'com.apple.mail-shared'
        key 'DisableURLLoading'
        value 1
        only_if { File.exist? "/Users/#{node['harden_os']['user']}/Library"\
        "/Preferences/com.apple.mail-shared.plist" }
      end
    end

    def harden_macos_devices
      puts("\nHardening Library: " + __method__.to_s)

      # TODO: Verify the value returned for the UIDFilter does not equal none
      # ^ Actually...it does. Why is that a problem? 
      mac_os_x_userdefaults 'Disable the IR controller (Apple Remote)' do
        domain '/Library/Preferences/com.apple.driver.AppleIRController'
        key 'DeviceEnabled'
        value 0
        only_if { File.exist? '/Library/Preferences/com.apple.driver.AppleIRController.plist' }
      end

      bluetooth_powerstate = Mixlib::ShellOut.new('defaults read /Library/Preferences/com.apple.Bluetooth ControllerPowerState | grep 1')
      if bluetooth_powerstate.run_command
        # Turn off Bluetooth, but only if no paired devices exist
        mac_os_x_userdefaults 'Turn off Bluetooth' do
          domain '/Library/Preferences/com.apple.Bluetooth'
          key 'ControllerPowerState'
          value 0
          notifies :run, 'execute[Kill bluetooth server process]', :delayed
          not_if 'system_profiler SPBluetoothDataType | grep "^Bluetooth:" -A 20 | grep "Connectable: Yes"'
        end
      end

      execute 'Kill bluetooth server process' do
        command 'killall -HUP blued'
        action :nothing
      end

      # Disables the ability to login via Smartcard (macOS 10.12)
      mac_os_x_userdefaults 'disable Sierra smartcard support' do
        domain '/Library/Preferences/com.apple.security.smartcard'
        key 'DisabledTokens'
        value 'com.apple.CryptoTokenKit.pivtoken'
        type 'array'
        only_if { node['platform_version'].split('.')[1].to_i >= 12 }
      end
    end

    def harden_macos_finder
      puts("\nHardening Library: " + __method__.to_s)

      harden_macos_atypical_plist 'Change spotlight enabled categories' do
        identifier 'Spotlight'
        changelist node['harden_os']['spotlight_plist_changelist']
        action :update
      end
    end

    def harden_macos_icloud
      puts("\nHardening Library: " + __method__.to_s)
      # This is reserved for future use
    end

    def harden_macos_network
      puts("\nHardening Library: " + __method__.to_s)

      get_wo_network = Mixlib::ShellOut.new('sudo systemsetup getwakeonnetworkaccess')
      unless get_wo_network.run_command == \
             'Wake On Network Access: Not supported on this machine.'
        execute 'Disable wake on network access' do
          command 'systemsetup -setwakeonnetworkaccess off'
          only_if { get_wo_network.run_command == 'Wake On Network Access: On' }
        end
      end

      # TODO - this command gives an error on macOS 10.11.  Check out.
      execute 'Require admin to create wifi computer-to-computer networks' do
        command '/System/Library/PrivateFrameworks/Apple80211.framework/'\
                'Versions/Current/Resources/airport en1 prefs'\
                'RequireAdminIBSS=YES'
        only_if '/System/Library/PrivateFrameworks/Apple80211.framework/'\
               'Versions/Current/Resources/airport en1 prefs RequireAdminIBSS'\
               ' | grep "RequireAdminIBSS=NO\|RequireAdminIBSS=Unknown"'
      end

      execute 'Set Clock Using Network Time' do
        command 'systemsetup setusingnetworktime on'
        only_if 'systemsetup getusingnetworktime | grep "Network Time: Off"'
      end
    end

    def harden_macos_remote_services
      puts("\nHardening Library: " + __method__.to_s)

      # Disable Apple Remote Events
      #   "Apple events are the message-based interprocess communication
      #   mechanism in Mac OS, first appearing in System 7 and supported by
      #   every version since then, including Mac OS X. Apple events describe
      #   "high-level" events such as "open document" or "print file," whereas
      #   earlier OSs had supported much more basic events, namely "click"
      #   and "keypress". Apple events form the basis of the Mac OS scripting
      #   system, AppleScript."
      execute 'Disable Apple Remote Events' do
        command 'systemsetup -setremoteappleevents off'
        only_if 'systemsetup -getremoteappleevents | grep'\
                '"Remote Apple Events: On"'
      end

      # Note: the agent does not start by default on 10.11
      execute 'Disable remote management' do
        command '/System/Library/CoreServices/RemoteManagement/ARDAgent.app'\
                '/Contents/Resources/kickstart -deactivate -stop -configure'\
                '-access -off'
        only_if 'ps -ef | egrep "/System/Library/CoreServices/Remote'\
                'Management/ARDAgent.app/Contents/MacOS/[A]RDAgent"'
      end
    end

    def harden_macos_sysprefs_sharing
      puts("\nHardening Library: " + __method__.to_s)

      # Most below are in GUI at System Preferences -> Sharing
      execute 'Disable remote login (SSH)' do
        command '/usr/sbin/systemsetup -f -setremotelogin off'
        only_if 'systemsetup -getremotelogin | grep "Remote Login: On"'
      end

      execute 'Disable printer sharing' do
        command 'cupsctl --no-share-printers'
        only_if 'cupsctl | egrep "_share_printers=1"'
      end

      macosx_service 'Disable screen sharing' do
        service_name 'com.apple.screensharing'
        plist '/System/Library/LaunchDaemons/com.apple.screensharing.plist'
        action :disable
      end

      # Disabled by default on MacOS >10.8
      macosx_service 'Disable ftp daemon' do
        service_name 'ftp'
        plist '/System/Library/LaunchDaemons/ftp.plist'
        action :disable
      end

      macosx_service 'Disable Apple File Server (File Sharing)' do
        service_name 'AppleFileServer'
        plist '/System/Library/LaunchDaemons/com.apple.AppleFileServer.plist'
        action :disable
      end

      macosx_service 'Disable SMB File Server (File Sharing)' do
        service_name 'smbd'
        plist '/System/Library/LaunchDaemons/com.apple.smbd.plist'
        action :disable
      end
    end

    def harden_macos_utilities
      puts("\nHardening Library: " + __method__.to_s)
      # This is reserved for future use.
    end

    def harden_macos_ssh_configs
      puts("\nHardening Library: " + __method__.to_s)

      # openSSH vulnerability fix - January 2016
      #   CVE-2016-0777 and CVE-2016-0778)
      #
      # The following variables must be set based on OS X version because the
      #   location of the ssh_config file changed between 10.10 and 10.11
      current_ssh_config = nil
      current_ssh_config = '/etc/ssh/ssh_config' if \
                            os_version == 11
      current_ssh_config = '/etc/ssh_config' if \
                            os_version == 10
      ssh_config_version = nil
      ssh_config_version = '/vulns/ssh_config1011' if \
                            os_version == 11
      ssh_config_version = '/vulns/ssh_config1010' if \
                            os_version == 10

      # Will not execute if osx_version is anything other than 10.10 or 10.11
      unless current_ssh_config.nil?
        require 'FileUtils'
        require 'tmpdir'

        # Create directory to hold new ssh config.  This must be done outside
        # a ruby_block so that it can be used in the cookbook_file below
        temp_dir = Dir.mktmpdir()
        puts(__method__.to_s + ": (2) temporary directory is: " + temp_dir)
        new_ssh_config_path = temp_dir + '/ssh_config'

        # Copy ssh_config from the cookbook and place into temporary directory
        cookbook_file "Secure SSH config file" do
          source ssh_config_version
          path new_ssh_config_path
          owner 'root'
          group 'wheel'
          action :create
        end

        # Rename existing ssh_config and then copy over the new config
        #   The existing (original) ssh_config has the day, time, and .bak
        #   appended to the file name. Only then does it copy over the new
        #   ssh_config file
        #
        ruby_block 'rename original ssh_config and copy over new config' do
          block do
            File.rename(current_ssh_config, current_ssh_config + Time.now.strftime("%Y-%m-%d_%H%M") + '.bak') unless current_ssh_config.empty?
            FileUtils.mv new_ssh_config_path, current_ssh_config unless current_ssh_config.empty?
          end
          not_if { FileUtils.compare_file(new_ssh_config_path, current_ssh_config) }
        end # ruby block

        # Clean up!
        ruby_block 'Clean up temp directory' do
          block do
            FileUtils.rm_r temp_dir 
          end
        end

      end #unless
    end # def
  end #module HardeningHelpers
end # module OSHardeningCookbookMacOS

Chef::Recipe.send(:include, OSHardeningCookbookMacOS::HardeningHelpers)
Chef::Resource.send(:include, OSHardeningCookbookMacOS::HardeningHelpers)
Chef::Provider.send(:include, OSHardeningCookbookMacOS::HardeningHelpers)
